### PR TITLE
Timestamp defaults

### LINKED ISSUE #1535: @default support for timestamps
Currently, the `@default` trait on `Timestamp` fields is not supported ([source](https://github.com/disneystreaming/smithy4s/blob/89e88c972174fa0a3b2853901a52a5a6a4662c6d/modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala#L1420))

smithy supports setting the default value that depend on `@timestampFormat`, it would be nice if the codegen supported that too.

I can look into it if it's useful to have this feature

### PR DESCRIPTION
**Pull request by msosnicki**
## PR Checklist (not all items are relevant to all PRs)

- [x] Added unit-tests (for runtime code)
- [x] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [ ] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [ ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [ ] Updated dynamic module to match generated-code behaviour
- [ ] Added documentation
- [x] Updated changelog

fixes #1535

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 1]
```diff
@@ -1377,61 +1393,89 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
+      case (node, Primitive.Blob) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Blob, Array.empty[Byte])
+      case (node, Primitive.Bool) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Bool, false)
+      case timestamp @ (node, Primitive.Timestamp) =>
+        val format = traits
+          .collectFirst { case t: TimestampFormatTrait =>
+            t.getFormat()
+          }
+          .getOrElse(TimestampFormatTrait.Format.DATE_TIME)
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: **issue:** the default should probably be `EPOCH_SECONDS`, it seems to be the case in e.g. AWS's protocol compliance tests.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 2]
```diff
@@ -1377,61 +1393,89 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
-    case (node, Primitive.Blob) if node == Node.nullNode =>
-      TypedNode.PrimitiveTN(Primitive.Blob, Array.empty[Byte])
-    case (node, Primitive.Bool) if node == Node.nullNode =>
-      TypedNode.PrimitiveTN(Primitive.Bool, false)
-    case (node, Primitive.Timestamp) if node == Node.nullNode =>
-      TypedNode.PrimitiveTN(
-        Primitive.Timestamp,
-        java.time.Instant.ofEpochSecond(0)
+      p: Primitive,
+      traits: List[Trait]
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: let's avoid passing the trait and just do something like : `parseEpoch(node).orElse(parseDateTime(node)).orElse(parseHttpDate(node))`

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ Thank you!
 # 0.18.21
```
Comment (kubukoz) [inline CHANGELOG.md:11]: thought: we're gonna have to move it to 0.18.22

Comment (lenguyenthanh) [reply CHANGELOG.md:11]: that seems correct, as `0.18.21` is already published

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 1 above)
Comment (msosnicki) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: Changed, for some reason I was sure that the `date-time` is default, but that's incorrect, just verified with encoders.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 2 above)
Comment (msosnicki) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: The `node` that is passed here lacks this information, it is just the value of the `@default` trait. Don't like the `traits` being passed particularly too, but wasn't sure what is the best way to pass the `@timestampFormat`, given that this is used in other places (and other types) too.

File: CHANGELOG.md (hunk 3 above)
Comment (msosnicki) [reply CHANGELOG.md:11]: Updated

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 2 above)
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: My point is that it's okay to not pass the timestamp format and just try all officially supported formats in sequence. I think that timestampFormat is probably the only trait that is taken into consideration by the canonical node representation of data in smithy, and given that the data is already validated by the smithy layer, it's fine to just assume that one of the formats we know is supported will be successful when decoding the value.

**Review by Baccata — APPROVED** [review]

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 1 above)
Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: it used to be :)

**Review by kubukoz — APPROVED** [review]

**Pull request merged**